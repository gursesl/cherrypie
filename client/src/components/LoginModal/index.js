/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Popup, Button, Header, Image, Modal, Grid, Message } from 'semantic-ui-react'
import { graphql } from 'react-apollo'
import { withRouter, Link } from 'react-router-dom'
import localforage from 'localforage'
import LoginForm from './loginForm'
import LoginSuccessModal from './loginSuccessModal'
import LoginFailureModal from './loginFailureModal'
import loginMutation from '../../graphql/mutations/loginMutation'
import usersListQuery from '../../graphql/queries/usersListQuery'
import currentUserQuery from '../../graphql/queries/currentUserQuery'
import logo from './logo.png'
// import findUserByEmailQuery from '../../graphql/queries/findUserByEmailQuery'

class LoginModal extends Component {
  state = {
    successModalOpen: false,
    failureModalOpen: false,
    next: null,
    errors: [],
  }

  componentWillMount() {
    console.log(`LoginModal:cwm: ${JSON.stringify(this.props)}`)
    if (this.props.location && this.props.location.state) {
      const { modalOpen, next } = this.props.location.state
      // const { next } = this.state
      this.setState({ modalOpen, next })
    }
  }

  // componentWillReceiveProps(nextProps) {
  //   if (nextProps.location && nextProps.location.state) {
  //     this.setState({ modalOpen: true })
  //   }
  // }

  saveToken = (key, value) => {
    localforage
      .setItem(key, value)
      .then(() => {
        // console.log(item)
      })
      .catch(() => {
        // console.log(err)
      })
  }

  handleOpen = () =>
    this.setState({
      modalOpen: true,
      successModalOpen: false,
      failureModalOpen: false,
    })

  handleClose = () =>
    this.setState({
      modalOpen: false,
      successModalOpen: false,
      failureModalOpen: false,
    })

  handleSuccess = () => {
    console.log('handleSuccess')
    console.log(this.props)
    console.log(this.state)
    // this.setState({ modalOpen: false, successModalOpen: true })
    // debugger
    this.setState({ modalOpen: false })
    if (this.state.next) {
      // this.context.history.push(this.state.next)
      // setTimeout(() => {
      this.props.history.push(this.state.next)
      // }, 10)
    }
  }

  handleFailure = errors => this.setState({ modalOpen: false, failureModalOpen: true, errors })
  handleLogin = (values) => {
    const variables = values.toJS()
    this.props
      .mutate({
        variables,
        // refetchQueries: [
        //   // { query: usersListQuery, variables },
        //   { query: currentUserQuery },
        //   // { query: findUserByEmailQuery, variables },
        // ],
        update: (proxy, { data: { loginUser } }) => {
          // Read the data from our cache for this query.
          const data = proxy.readQuery({ query: currentUserQuery })

          // Add our user from the mutation to the end
          data.getCurrentUser = loginUser

          // Write our data back to the cache
          proxy.writeQuery({ query: currentUserQuery, data })
        },
      })
      .then((response) => {
        const {
          ok, token, refreshToken, errors,
        } = response.data.loginUser
        if (ok) {
          this.saveToken('token', token)
          this.saveToken('refreshToken', refreshToken)
          this.handleSuccess()
        } else {
          this.handleFailure(errors)
        }
      })
      .catch((errors) => {
        if (errors.message) {
          this.handleFailure([{ message: errors.message, path: 'NetworkError' }])
        } else {
          this.handleFailure(errors.graphQLErrors)
        }
      })
  }

  render() {
    return (
      <div>
        <Popup trigger={<Button onClick={this.handleOpen}>Sign in</Button>}>
          <Popup.Header>Sign in details</Popup.Header>
          <Popup.Content>Click here to sign in</Popup.Content>
        </Popup>

        <Modal open={this.state.modalOpen} onClose={this.handleClose} basic size="small">
          <Modal.Content>
            <Grid textAlign="center" style={{ height: '100%' }} verticalAlign="middle">
              <Grid.Column style={{ maxWidth: 450 }}>
                <Header as="h2" color="teal" textAlign="center">
                  <Image src={logo} /> Sign in
                </Header>
                <LoginForm {...this.props} onSubmit={this.handleLogin} />
                <Message>
                  Don&#39;t have an account?{' '}
                  <Link
                    href="/register"
                    to={{
                      pathname: '/register',
                      state: { modalOpen: true },
                    }}
                  >
                    Sign up
                  </Link>
                </Message>
              </Grid.Column>
            </Grid>
          </Modal.Content>
        </Modal>
        <LoginSuccessModal
          successModalOpen={this.state.successModalOpen}
          handleClose={this.handleClose}
        />
        <LoginFailureModal
          successModalOpen={this.state.failureModalOpen}
          errors={this.state.errors}
          handleClose={this.handleClose}
        />
      </div>
    )
  }
}

LoginModal.propTypes = {
  mutate: PropTypes.func.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      modalOpen: PropTypes.bool,
    }),
  }),
}

LoginModal.defaultProps = {
  location: {
    state: {
      modalOpen: false,
    },
  },
}

export default withRouter(graphql(loginMutation)(LoginModal))
