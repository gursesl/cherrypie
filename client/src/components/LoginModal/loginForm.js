import React from 'react'
import PropTypes from 'prop-types'
import { Button, Form, Segment, Message } from 'semantic-ui-react'
import styled from 'styled-components'
import { Field, reduxForm } from 'redux-form/immutable'
import validator from 'validator'
import client from '../../apolloClient'
import findUserByEmailQuery from '../../graphql/queries/findUserByEmailQuery'

const StyledInputDiv = styled.div`margin: 10px 0;`

const emailField = field => (
  <StyledInputDiv>
    <Form.Input
      {...field.input}
      icon="envelope"
      iconPosition="left"
      placeholder="Email address"
      loading={field.meta.asyncValidating}
    />
    {field.meta.touched &&
      field.meta.error && (
        <Message size="mini" negative visible={false}>
          <Message.Header>{field.meta.error}</Message.Header>
        </Message>
      )}
  </StyledInputDiv>
)

const passwordField = field => (
  <StyledInputDiv>
    <Form.Input
      {...field.input}
      fluid
      icon="lock"
      iconPosition="left"
      placeholder="Password"
      type="password"
    />
    {field.meta.touched &&
      field.meta.error && (
        <Message size="mini" negative visible={false}>
          <Message.Header>{field.meta.error}</Message.Header>
        </Message>
      )}
  </StyledInputDiv>
)

const validate = (values) => {
  const errors = {}
  if (!values.get('email') || !validator.isEmail(values.get('email'))) {
    errors.email = 'A valid email address is required.'
  }
  if (
    !values.get('password') ||
    !validator.isLength(values.get('password'), { min: 3, max: 100 })
  ) {
    errors.password = 'A valid password between 3-100 characters is required.'
  }
  return errors
}

const asyncValidate = values =>
  client
    .query({
      query: findUserByEmailQuery,
      variables: {
        email: values.get('email'),
      },
    })
    .then((response) => {
      // console.log(response)
      if (!response.data.findUserByEmail) {
        throw { email: `${values.get('email')} is not registered.` } //eslint-disable-line
      }
    })

const RegisterForm = (props) => {
  const { handleSubmit, pristine, submitting } = props

  return (
    <Segment padded>
      <Form onSubmit={handleSubmit}>
        <Field name="email" component={emailField} />
        <Field name="password" component={passwordField} />
        <Button color="teal" fluid size="large" type="submit" disabled={pristine || submitting}>
          Sign in
        </Button>
      </Form>
    </Segment>
  )
}

RegisterForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  pristine: PropTypes.bool.isRequired,
  submitting: PropTypes.bool.isRequired,
}

export default reduxForm({
  form: 'loginForm',
  validate,
  asyncValidate,
  asyncBlurFields: ['email'],
})(RegisterForm)
