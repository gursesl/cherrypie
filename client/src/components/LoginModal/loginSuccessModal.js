import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button, Header, Modal, Icon } from 'semantic-ui-react'

class LoginSuccessModal extends Component {
  state = {
    modalOpen: false,
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      modalOpen: nextProps.successModalOpen,
    })
  }

  handleOpen = () => this.setState({ modalOpen: true })
  handleClose = () => this.setState({ modalOpen: false })

  render() {
    return (
      <Modal open={this.state.modalOpen} onClose={this.props.handleClose} basic size="small">
        <Header icon="user" content="Hooray!" />
        <Modal.Content>
          <h3>You have been logged in succcessfully.</h3>
        </Modal.Content>
        <Modal.Actions>
          <Button color="green" onClick={this.props.handleClose} inverted>
            <Icon name="checkmark" /> Start
          </Button>
        </Modal.Actions>
      </Modal>
    )
  }
}

LoginSuccessModal.propTypes = {
  successModalOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
}

export default LoginSuccessModal
