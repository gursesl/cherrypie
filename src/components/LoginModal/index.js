import React, { Component } from 'react'
import { Popup, Button, Modal, Form, Checkbox } from 'semantic-ui-react'

class LoginModal extends Component {
  state = { open: false }

  show = dimmer => () => this.setState({ dimmer, open: true })
  close = () => this.setState({ open: false })
  success = () => {
    this.setState({ open: false })
  }

  render() {
    const { open, dimmer } = this.state

    return (
      <div>
        <Popup trigger={<Button color="teal" onClick={this.show(true)}>Login</Button>}>
          <Popup.Header>Login details</Popup.Header>
          <Popup.Content>
            Use your email or username to login to Cherrypie
          </Popup.Content>
        </Popup>

        <Modal dimmer={dimmer} open={open} onClose={this.close}>
          <Modal.Header>Login</Modal.Header>
          <Modal.Content image>
            <Modal.Description>
              <Form>
                <Form.Input label="Username" type="text" />
                <Form.Input label="Password" type="password" />
                <Form.Field>
                  <Checkbox label="Remember me for 30 days" />
                </Form.Field>
              </Form>
            </Modal.Description>
          </Modal.Content>
          <Modal.Actions>
            <Button basic color="red" onClick={this.close}>Cancel</Button>
            <Button positive icon="sign in" labelPosition="right" content="Login" onClick={this.success} />
          </Modal.Actions>
        </Modal>
      </div>
    )
  }
}

export default LoginModal
