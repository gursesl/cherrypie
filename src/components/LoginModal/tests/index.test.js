import React from 'react'
import { shallow } from 'enzyme'
import { Button } from 'semantic-ui-react'
import '../../../setupTests'
import LoginModal from '..'

const component = shallow(<LoginModal />)

describe('LoginModal:index', () => {
  it('should be available', () => {
    expect(component.state().open).toBeFalsy()
    expect(component.find('Modal').length).toBe(1)
    expect(component.find('Button').length).toBe(2)
    expect(component).toMatchSnapshot()
  })

  it('should hover correctly', () => {
    const button = component.find('Popup').dive().find('Portal')
    const buttonType = button.props().trigger.type
    expect(buttonType).toBe(Button)

    button.simulate('onMouseOver')
    // Hover delayed
    setTimeout(() => {
      expect(component.state().open).toBeTruthy()
    }, 0)

    button.simulate('onMouseOut')
    // Hover delayed
    setTimeout(() => {
      expect(component.state().open).toBeFalsy()
    }, 0)
  })

  // it('should have a correct show method implementation', () => {
  //   expect(component.state().open).toBeFalsy()
  //   component.instance().show(true, )

  //   setTimeout(() => {
  //     expect(component.state().open).toBeTruthy()
  //   }, 0)
  // })

  it('should be visible when the Login button is clicked', () => {
    const loginButton = component.find('Popup').dive().find('Portal')
    const modal = component.find('Modal')
    expect(modal.props().open).toBeFalsy()

    loginButton.simulate('click')
    // Click delayed
    setTimeout(() => {
      expect(modal.props().open).toBeTruthy()
    }, 0)
  })

  it('should be hidden when the Cancel button is clicked', () => {
    const modal = component.find('Modal')
    const cancelButton = component.find('Modal').find('Button').at(0)
    expect(modal.props().open).toBeFalsy()

    cancelButton.simulate('click')
    // Click delayed
    setTimeout(() => {
      expect(modal.props().open).toBeFalsy()
    }, 0)
  })

  it('should be hidden when the form Login button is clicked', () => {
    const modal = component.find('Modal')
    const formLoginButton = component.find('Modal').find('Button').at(1)
    expect(modal.props().open).toBeFalsy()

    formLoginButton.simulate('click')
    // Click delayed
    setTimeout(() => {
      expect(modal.props().open).toBeFalsy()
    }, 0)
  })
})
