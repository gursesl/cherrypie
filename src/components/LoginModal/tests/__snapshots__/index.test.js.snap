// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModal:index should be available 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginModal />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Popup
          on="hover"
          position="top left"
          trigger={
            <Button
              as="button"
              color="teal"
              onClick={[Function]}
            >
              Login
            </Button>
          }
        >
          <PopupHeader>
            Login details
          </PopupHeader>
          <PopupContent>
            Use your email or username to login to Cherrypie
          </PopupContent>
        </Popup>,
        <Modal
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          onClose={[Function]}
          open={false}
        >
          <ModalHeader>
            Login
          </ModalHeader>
          <ModalContent
            image={true}
          >
            <ModalDescription>
              <Form
                as="form"
              >
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  label="Username"
                  type="text"
                />
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  label="Password"
                  type="password"
                />
                <FormField>
                  <Checkbox
                    label="Remember me for 30 days"
                    type="checkbox"
                  />
                </FormField>
              </Form>
            </ModalDescription>
          </ModalContent>
          <ModalActions>
            <Button
              as="button"
              basic={true}
              color="red"
              onClick={[Function]}
            >
              Cancel
            </Button>
            <Button
              as="button"
              content="Login"
              icon="sign in"
              labelPosition="right"
              onClick={[Function]}
              positive={true}
            />
          </ModalActions>
        </Modal>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <PopupHeader>
              Login details
            </PopupHeader>,
            <PopupContent>
              Use your email or username to login to Cherrypie
            </PopupContent>,
          ],
          "on": "hover",
          "position": "top left",
          "trigger": <Button
            as="button"
            color="teal"
            onClick={[Function]}
          >
            Login
          </Button>,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "children": "Login details",
            },
            "ref": null,
            "rendered": "Login details",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "children": "Use your email or username to login to Cherrypie",
            },
            "ref": null,
            "rendered": "Use your email or username to login to Cherrypie",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <ModalHeader>
              Login
            </ModalHeader>,
            <ModalContent
              image={true}
            >
              <ModalDescription>
                <Form
                  as="form"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Username"
                    type="text"
                  />
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Password"
                    type="password"
                  />
                  <FormField>
                    <Checkbox
                      label="Remember me for 30 days"
                      type="checkbox"
                    />
                  </FormField>
                </Form>
              </ModalDescription>
            </ModalContent>,
            <ModalActions>
              <Button
                as="button"
                basic={true}
                color="red"
                onClick={[Function]}
              >
                Cancel
              </Button>
              <Button
                as="button"
                content="Login"
                icon="sign in"
                labelPosition="right"
                onClick={[Function]}
                positive={true}
              />
            </ModalActions>,
          ],
          "closeOnDimmerClick": true,
          "closeOnDocumentClick": false,
          "dimmer": true,
          "eventPool": "Modal",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "children": <ModalDescription>
                <Form
                  as="form"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Username"
                    type="text"
                  />
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Password"
                    type="password"
                  />
                  <FormField>
                    <Checkbox
                      label="Remember me for 30 days"
                      type="checkbox"
                    />
                  </FormField>
                </Form>
              </ModalDescription>,
              "image": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": <Form
                  as="form"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Username"
                    type="text"
                  />
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Password"
                    type="password"
                  />
                  <FormField>
                    <Checkbox
                      label="Remember me for 30 days"
                      type="checkbox"
                    />
                  </FormField>
                </Form>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {
                  "as": "form",
                  "children": Array [
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Username"
                      type="text"
                    />,
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Password"
                      type="password"
                    />,
                    <FormField>
                      <Checkbox
                        label="Remember me for 30 days"
                        type="checkbox"
                      />
                    </FormField>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "as": [Function],
                      "control": [Function],
                      "label": "Username",
                      "type": "text",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "as": [Function],
                      "control": [Function],
                      "label": "Password",
                      "type": "password",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "children": <Checkbox
                        label="Remember me for 30 days"
                        type="checkbox"
                      />,
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": null,
                      "nodeType": "class",
                      "props": Object {
                        "label": "Remember me for 30 days",
                        "type": "checkbox",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Button
                  as="button"
                  basic={true}
                  color="red"
                  onClick={[Function]}
                >
                  Cancel
                </Button>,
                <Button
                  as="button"
                  content="Login"
                  icon="sign in"
                  labelPosition="right"
                  onClick={[Function]}
                  positive={true}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "basic": true,
                  "children": "Cancel",
                  "color": "red",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Cancel",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "content": "Login",
                  "icon": "sign in",
                  "labelPosition": "right",
                  "onClick": [Function],
                  "positive": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Popup
            on="hover"
            position="top left"
            trigger={
              <Button
                as="button"
                color="teal"
                onClick={[Function]}
              >
                Login
              </Button>
            }
          >
            <PopupHeader>
              Login details
            </PopupHeader>
            <PopupContent>
              Use your email or username to login to Cherrypie
            </PopupContent>
          </Popup>,
          <Modal
            closeOnDimmerClick={true}
            closeOnDocumentClick={false}
            dimmer={true}
            eventPool="Modal"
            onClose={[Function]}
            open={false}
          >
            <ModalHeader>
              Login
            </ModalHeader>
            <ModalContent
              image={true}
            >
              <ModalDescription>
                <Form
                  as="form"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Username"
                    type="text"
                  />
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    label="Password"
                    type="password"
                  />
                  <FormField>
                    <Checkbox
                      label="Remember me for 30 days"
                      type="checkbox"
                    />
                  </FormField>
                </Form>
              </ModalDescription>
            </ModalContent>
            <ModalActions>
              <Button
                as="button"
                basic={true}
                color="red"
                onClick={[Function]}
              >
                Cancel
              </Button>
              <Button
                as="button"
                content="Login"
                icon="sign in"
                labelPosition="right"
                onClick={[Function]}
                positive={true}
              />
            </ModalActions>
          </Modal>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <PopupHeader>
                Login details
              </PopupHeader>,
              <PopupContent>
                Use your email or username to login to Cherrypie
              </PopupContent>,
            ],
            "on": "hover",
            "position": "top left",
            "trigger": <Button
              as="button"
              color="teal"
              onClick={[Function]}
            >
              Login
            </Button>,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": "Login details",
              },
              "ref": null,
              "rendered": "Login details",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": "Use your email or username to login to Cherrypie",
              },
              "ref": null,
              "rendered": "Use your email or username to login to Cherrypie",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <ModalHeader>
                Login
              </ModalHeader>,
              <ModalContent
                image={true}
              >
                <ModalDescription>
                  <Form
                    as="form"
                  >
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Username"
                      type="text"
                    />
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Password"
                      type="password"
                    />
                    <FormField>
                      <Checkbox
                        label="Remember me for 30 days"
                        type="checkbox"
                      />
                    </FormField>
                  </Form>
                </ModalDescription>
              </ModalContent>,
              <ModalActions>
                <Button
                  as="button"
                  basic={true}
                  color="red"
                  onClick={[Function]}
                >
                  Cancel
                </Button>
                <Button
                  as="button"
                  content="Login"
                  icon="sign in"
                  labelPosition="right"
                  onClick={[Function]}
                  positive={true}
                />
              </ModalActions>,
            ],
            "closeOnDimmerClick": true,
            "closeOnDocumentClick": false,
            "dimmer": true,
            "eventPool": "Modal",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": "Login",
              },
              "ref": null,
              "rendered": "Login",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "children": <ModalDescription>
                  <Form
                    as="form"
                  >
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Username"
                      type="text"
                    />
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Password"
                      type="password"
                    />
                    <FormField>
                      <Checkbox
                        label="Remember me for 30 days"
                        type="checkbox"
                      />
                    </FormField>
                  </Form>
                </ModalDescription>,
                "image": true,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "children": <Form
                    as="form"
                  >
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Username"
                      type="text"
                    />
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      label="Password"
                      type="password"
                    />
                    <FormField>
                      <Checkbox
                        label="Remember me for 30 days"
                        type="checkbox"
                      />
                    </FormField>
                  </Form>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {
                    "as": "form",
                    "children": Array [
                      <FormInput
                        as={[Function]}
                        control={[Function]}
                        label="Username"
                        type="text"
                      />,
                      <FormInput
                        as={[Function]}
                        control={[Function]}
                        label="Password"
                        type="password"
                      />,
                      <FormField>
                        <Checkbox
                          label="Remember me for 30 days"
                          type="checkbox"
                        />
                      </FormField>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": null,
                      "nodeType": "function",
                      "props": Object {
                        "as": [Function],
                        "control": [Function],
                        "label": "Username",
                        "type": "text",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": null,
                      "nodeType": "function",
                      "props": Object {
                        "as": [Function],
                        "control": [Function],
                        "label": "Password",
                        "type": "password",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": null,
                      "nodeType": "function",
                      "props": Object {
                        "children": <Checkbox
                          label="Remember me for 30 days"
                          type="checkbox"
                        />,
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": null,
                        "nodeType": "class",
                        "props": Object {
                          "label": "Remember me for 30 days",
                          "type": "checkbox",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": [Function],
                      },
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Button
                    as="button"
                    basic={true}
                    color="red"
                    onClick={[Function]}
                  >
                    Cancel
                  </Button>,
                  <Button
                    as="button"
                    content="Login"
                    icon="sign in"
                    labelPosition="right"
                    onClick={[Function]}
                    positive={true}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "basic": true,
                    "children": "Cancel",
                    "color": "red",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Cancel",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "content": "Login",
                    "icon": "sign in",
                    "labelPosition": "right",
                    "onClick": [Function],
                    "positive": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
